modules/database/database.bal

```diff
@@ -959,44 +957,155 @@ public isolated transactional function createTimelineSubstep(int stepId, Timelin
             );
-            return;
         }
     }
-    check createTimelineSubstep(stepId, substep);
+    if result?.affectedRowCount != 1 {
+        int? rowCount = result?.affectedRowCount;
+        return error(string `Update not successful. Affected rows: ${rowCount != () ? rowCount : 0}`);
+    }
 }
 
-public isolated transactional function updateTimelineProgress(int stepId, Progress progress) returns error? {
-    var insertResult = check experimentDB->execute(
-        `UPDATE TimelineStep SET pStart = ${progress.progressStart}, pTarget = ${progress.progressTarget}, pValue = ${progress.progressValue}, pUnit = ${progress.progressUnit} WHERE stepId = ${stepId};`
-    );
+# Updates database from a list of received substeps. First checks received list against old list of substeps for the given step in the database to find updated and new substeps. If steps are missing or illegal changes were made (apart from setting cleared to 1) or multiple new uncleared substeps are added an appropriate error is returned. If at least one new substep is added, all old substeps are automatically cleared (in case they are not cleared in the received list). A warning is printed if more than one new substep is added. 
+#
+# + stepId - Timeline step id
+# + receivedSubsteps - list of received substeps
+# + return - Returns true if changes were made successfully and false if no changes were made. Else raises an error.
+public isolated transactional function updateTimelineSubsteps(int stepId, TimelineSubstep[] receivedSubsteps) returns boolean|error {
+    TimelineSubstepSQL[] tempDBSubsteps = check getTimelineSubsteps(stepId);
+    // cast db substeps to TimelineSubstep
+    TimelineSubstep[] oldDBSubsteps = [];
+    foreach var tempDBsubstep in tempDBSubsteps {
+        TimelineSubstep tempDBSubstep = {
+            substepId: tempDBsubstep.substepId,
+            href: tempDBsubstep.href,
+            hrefUi: tempDBsubstep.hrefUi,
+            cleared: tempDBsubstep.cleared
+        };
+        oldDBSubsteps.push(tempDBSubstep);
+    }
+
+    TimelineSubstep[] newSubsteps = [];
+    TimelineSubstep[] updatedSubsteps = [];
+    TimelineSubstep[] oldUnclearedSubsteps = [];
+    TimelineSubstep oldDBSubstep;
+
+    // find new/changed substeps
+    int oldOrUpdatedCounter = 0;
+    int unclearedCounter = 0;
+    foreach var receivedSubstep in receivedSubsteps {
+        if receivedSubstep.cleared == 0 {
+            unclearedCounter += 1;
+        }
+        if oldDBSubsteps.indexOf(receivedSubstep) != () {
+            oldOrUpdatedCounter += 1;
+            if receivedSubstep.cleared == 0 {
+                oldUnclearedSubsteps.push(receivedSubstep);
+            }
+        } else {
+            TimelineSubstep[] tmp = from var tempSubstep in oldDBSubsteps
+                where tempSubstep.href == receivedSubstep.href
+                select tempSubstep; // hrefs must be unique in each step
```

Comment:
In this case I would suggest to refactor the substep data model to include the substep number. Otherwise, we have no way to distinguish between two substeps with the same href (see above).

Reply:
See my comment above. The substepNr is already there as it is the list index. Though it might be a good idea to explicitly extract this information into the substep record.